### ISSUE TITLE
ValueError: too many file descriptors in select()

### ISSUE THREAD

**Issue by uFire** [spesmilo/electrum#5873]
Hi 

I have 3-5 instances of Electrum (3.3.8, BTC) with hundreds of transactions in each open for days and weeks on my Win10 x64 PC and sometimes they crash with the following error: 

```
20200101T155314.140910Z |    ERROR | gui.qt.exception_window.Exception_Hook | exception caught by crash reporter
Traceback (most recent call last):
  File "electrum\util.py", line 879, in run_with_except_hook
  File "threading.py", line 864, in run
  File "asyncio\base_events.py", line 471, in run_until_complete
  File "asyncio\base_events.py", line 438, in run_forever
  File "asyncio\base_events.py", line 1415, in _run_once
  File "selectors.py", line 323, in select
  File "selectors.py", line 314, in _select
ValueError: too many file descriptors in select()
```

Tell me if you need any additional info to fix that.

**Comment by SomberNight** [comment]
If it only happens after running for a long time, maybe we are not closing sockets properly... ?

[Here on stackoverflow](https://stackoverflow.com/a/47689451) it is suggested we use [asyncio.ProactorEventLoop](https://docs.python.org/3.8/library/asyncio-eventloop.html#asyncio.ProactorEventLoop) as the event loop implementation on Windows.
We currently require Python 3.6, and ship Python 3.6 with the Windows binaries.
In Python 3.8, `ProactorEventLoop` was made the default event loop implementation on Windows.
Seems to me that `ProactorEventLoop` is "production-ready" since Python 3.5 (that was when TLS support was added).

I guess we could explicitly use `ProactorEventLoop` instead of `SelectorEventLoop` on Windows, e.g.:

```patch
diff --git a/electrum/util.py b/electrum/util.py
index c946c53c9..b91720e34 100644
--- a/electrum/util.py
+++ b/electrum/util.py
@@ -1144,6 +1144,9 @@ def create_and_start_event_loop() -> Tuple[asyncio.AbstractEventLoop,
             return
         loop.default_exception_handler(context)
 
+    if sys.platform == 'win32':
+        loop = asyncio.ProactorEventLoop()
+        asyncio.set_event_loop(loop)
     loop = asyncio.get_event_loop()
     loop.set_exception_handler(on_exception)
     # loop.set_debug(1)
```
TBH I don't really understand the implications of a change like this.

@ysangkok Any thoughts?

**Comment by SomberNight** [comment]
@uFire 
> I have 3-5 instances of Electrum (3.3.8, BTC) with hundreds of transactions in each open for days and weeks on my Win10 x64 PC

Do you mean you have 3-5 wallets open in a single python process, using the Qt GUI, on a single PC?
Or do you run Electrum in daemon mode? Maybe multiple processes/daemons?

**Comment by uFire** [comment]
@SomberNight , sorry, probably my English is not good enough to explain what I mean. Yes, you right - I have a single instance of electrum 3.3.8 installed and always have open 3-5 wallets (3 most of the time) on a single PC. And they run for weeks 24/7. They are connected to my own local ElectrumX 1.13.0 server.

Also I got the same error in the past using other Win10 x64 laptop - electrum wallet windows just stop updating transactions/exchange rate and give me a bug reporting window with the error I posted in the first post.

**Comment by uFire** [comment]
Hello

I am keep getting that error every 2-3 days. Is there any workaround you could suggest?

Also, the Electrum's built-in bug reporting under Windows seems to not working. After I click "Report now", there is a small window saying "Sending bug report" and it stay infinite.

Thanks :)

```
Traceback
Traceback (most recent call last):
  File "electrum\util.py", line 879, in run_with_except_hook
  File "threading.py", line 864, in run
  File "asyncio\base_events.py", line 471, in run_until_complete
  File "asyncio\base_events.py", line 438, in run_forever
  File "asyncio\base_events.py", line 1415, in _run_once
  File "selectors.py", line 323, in select
  File "selectors.py", line 314, in _select
ValueError: too many file descriptors in select()
 
Additional information
Electrum version: 3.3.8 
Python version: 3.6.8 (tags/v3.6.8:3c6b436a57, Dec 23 2018, 23:31:17) [MSC v.1916 32 bit (Intel)] 
Operating system: Windows-10-10.0.18362-SP0 
Wallet type: standard 
Locale: en_GB
```

**Comment by ysangkok** [comment]
@SomberNight I have been looking for issues with ProactorEventLoop, here are a few potential issues:

* No dedicated policy is available before 3.8, I don't know if that is a problem: https://github.com/python/cpython/commit/8f4042964d5b0ddf5cdf87862db962ba64e3f64a
* Signal handlers not supported correctly before 3.8: https://bugs.python.org/issue23057
* UDP not supported before 3.8: https://github.com/python/cpython/commit/bafd4b5ac83b6cc0b7455290a04c4bfad34bdc90
* Assertion error which might still affect 3.6: https://bugs.python.org/issue33833

**Comment by SomberNight** [comment]
We could bump the python version in the Windows binaries to 3.8 when PyInstaller 4.1 gets released, as they have added support for it now: https://github.com/pyinstaller/pyinstaller/issues/4311

Meanwhile Electrum 4.0.3 bundles Python 3.7.9.
@uFire I wonder if you can reproduce this issue with Electrum 4.0.3.

**Comment by uFire** [comment]
@SomberNight Thank you for remembering about my issue :) I even forgot about this thread.

The issue is gone after I upgraded to 4.0.2. Now I am on 4.0.3 and everything is seems to be fine. I am running four instances always open 24/7 making hundreds of transactions daily and there is no single crash since 4.0.2.

**Comment by SomberNight** [comment]
So maybe this was a bug in Python and it got fixed between 3.6.8 and 3.7.9.
